Enhancement: The encoder, but cleaner this time (hopefully)

haltman-at opened the pull request:
The contents of this PR are identical to #3974, so **please see that for a description**, including questions I would like resolved.

What's different between this PR and #3974 is that I've reorganized it into just 6 commits, to make reviewing it easier.  Hope this is helpful @gnidan!

Note that I have not yet made any of the changes I proposed in #3974, since I was waiting for review first.  Also uh I guess I never really fixed any typedoc problems, I (or rather @gnidan) just got it to the point where the typedoc compiles, but I just remembered I never actually bothered checking out the results and fixing problems there, so I may still need to do that...

Edit: Addresses issues #2189 and #3819

gnidan reviewed (changes requested):
This review is for just the first commit, since there's likely a lot of discussion to be had even just with the codec changes. Mostly, I'm very keen on getting rid of the long series of nested `if` statements everywhere. Let's do away with that pattern.

gnidan commented on packages/codec/lib/wrap/elementary.ts:
```diff
@@ -0,0 +1,1254 @@
+  wrapOptions: WrapOptions
+): Format.Values.BoolValue {
+  let asBoolean: boolean;
+  if (typeof input === "boolean") {
```
Alternatively, maybe we don't want to separate `condition` and `transform` and instead do something like this:
```typescript
export type MakeWrapCase<Input, Output> = (input: unknown) => void | { match: Output };

const wrapBool: Wrap<Format.Types.Type, Format.Values.Value> = makeWrap({
  cases: cases<boolean>()
    .add(input => typeof input === "boolean" ? { match: input } : undefined)
    .add(input => input === "true" || input === "false" ? { match: input === "true" } : undefined)
    .add(input => typeof input === "number" ? { match: input !== 0 } : undefined),
  process: (asBoolean, dataType) => ({
    type: dataType,
    value: {
      asBoolean
    }
  })
});
```

haltman-at replied:
I ended up doing something vaguely along these lines, marking this resolved.

haltman-at commented on packages/codec/lib/wrap/elementary.ts:
```diff
@@ -0,0 +1,1254 @@
+
+export function wrapBool(
+  dataType: Format.Types.BoolType,
+  input: any,
```
OK, I went and changed these.  In one or two cases the necessary adjustment is a little hacky, but for the most part no adjustment was needed beyond simply changing `any` to `unknown`.

haltman-at commented on packages/codec/lib/format/values.ts:
```diff
@@ -58,7 +58,9 @@ export type Result =
   | FunctionExternalResult
-  | FunctionInternalResult;
+  | FunctionInternalResult
+  | OptionsResult;
```
That's there to support options in overload resolution.  I'll skip going into further detail here since we discussed this elsewhere.

haltman-at commented on packages/codec/lib/common/types.ts:
```diff
@@ -51,3 +53,46 @@ export class UnknownUserDefinedTypeError extends Error {
   }
 }
+
+/**
+ * Type for transaction options, including
+ * Quorum-specific ones (privateFor) and
+ * Truffle-specific ones (overwrite)
+ *
+ * @Category Interfaces
+ */
+export interface Options {
```
I don't see the need, outside of encoder purposes?

gnidan replied:
Sure, that's fine

haltman-at commented:
Heh, I only just noticed that maybe `watchMappingKey` / `constructSlot` should also accept `String`s in place of `string`s when specifying struct members...

haltman-at commented on packages/codec/lib/wrap/elementary.ts:
```diff
@@ -0,0 +1,1254 @@
+  dataType: Format.Types.BoolType,
+  input: any,
+  wrapOptions: WrapOptions
+): Format.Values.BoolValue {
```
OK, I did something like this with the `Case` type!

haltman-at commented:
Oh, while I was rewriting it, I also made some functional changes.

1. Nested type/value input is now disallowed (something that was impossible before!).
2. When `loose` is turned on, tuples and options will allow for the possibility that a `type/value` input or a wrapped result input should actually be treated as a plain object.
3. (Bug fix) The `loose` option previously was turned on, well, basically always, which is not how it was supposed to work.  Now it's only turned on when serious overload resolution comes into play.  (Not number-of-arguments-based overload resolution.)

Also, I added additional tests to cover these changes.

gnidan commented on packages/encoder/lib/encoders.ts:
```diff
@@ -0,0 +1,1302 @@
+   *
+   * See [[wrapForTransaction]] for documentation of the inputs.
+   */
+  public async encodeTransaction(
```
As discussed, I think this `encodeTransaction` method should be considered the primary interface - likely we want to `s/resolveAndEncode/encodeTransaction` (renaming this method to something else).

As part of this, we ought to accept function name strings in addition to `Method` objects directly.

haltman-at replied:
OK, I added `forAddress`.  I haven't tested it though.  Also, it's largely copypasted from the decoder.  It doesn't accept a block parameter though, it just uses `"latest"`.

I also haven't done the renaming yet.

haltman-at replied:
OK, did that renaming!  @gnidan, still waiting for answers about everything else.

gnidan replied:
> I noticed that the Decoder, if you try to spawn a ContractDecoder by name, but it has no bytecode, it'll do a lookup by name. D: Yikes! And of course the Encoder now copies that behavior... are we OK with that for now, or does that need changing?

I say that this probably needs changing. Lookup by name is blurgh, too risky.

> Note: This opens the question of whether we want a special error for when the user puts in a string that doesn't actually match any function name.

Seems reasonable!

> OK, I added `forAddress`. I haven't tested it though. Also, it's largely copypasted from the decoder. It doesn't accept a block parameter though, it just uses "latest".

Any way to make it accept a block number?

(Any other points I missed here?)

haltman-at replied:
> I say that this probably needs changing. Lookup by name is blurgh, too risky.

Yeah, still need to figure out what to do about this.  We'll talk more about this I guess.

> Seems reasonable!

Fortunately, I've already done this! :)

> Any way to make it accept a block number?

IIRC, the reason I didn't do so is because the adapter function for getting code is typed as taking a regularized block specifier, and I didn't want to drag in the whole regularization mechanism here.  However, I *already* just hackily changed that type, so... yeah, why not, I guess I can.  Of course it'll have to be a block *number* (or one of the specific allowed strings), rather than a block hash, but I guess that's already how decoder works.

haltman-at replied:
OK, I've changed it to allow an optional block number.

gnidan replied:
Re: project encoder concerns, let's go with a stopgap solution of marking this method internal-only.

haltman-at replied:
OK, I hid those project encoder methods, and adjusted other documentation appropriately.

While I was at it, I realized that not all of the things that ought to allow taking a function name did, so I made that possible, and factored the code that does it.

I'm realizing now though that the documentation is written in a way that doesn't really focus things around `encodeTransaction` like they probably should, but I'll deal with that later?

haltman-at replied:
Also I still need to deal with the name lookup problem.

haltman-at replied:
OK, after discussing things with @gnidan, I changed things so that it will still look up by name if there's a unique name match (note this is only if there's no bytecode match of course!), but otherwise, rather than failing, will return a minimal default result.  However, in all cases, we then trust the passed-in ABI rather than the found ABI.

Note I didn't add any tests of the default result case.

gnidan reviewed (commented):
Alright Harry here are the promised "quick comments" :)

haltman-at commented:
Hm, I found a problem with the error specificity system.  Errors for individual components (like of an array, say) weren't being treated as specific enough relative to errors in the form of the whole.  So I added a "specificity floor" setting; when this is set, *after* the best error is selected, it will have its specificity promoted to be at least the floor, if it wasn't already.  Again, this happens *after* the best error is selected, so it doesn't erase comparisons between errors in the *same* call of `wrapWithCases`; it only promotes the specificities of errors in the components relative to the specificity of errors in the whole.

haltman-at commented:
Added specific error for if attempting to resolve by name and there's no function by that name.

haltman-at commented:
OK, I've expanded UDVT support.

Note: If you're wondering why only bools, and not integers or decimals, have a UDVT error input case, it's because you can't make UDVTs on top of enums.

haltman-at commented:
OK, I have now finally implemented the loose bytes thing, and added tests for it.

Ultimately, I ended up restricting this to the dynamic case.  The problem is that if I implemented it for the static case, it could potentially lead to input being interpreted in some really unexpected ways.  I figured we didn't want that to happen; fortunately, ethers only allows it in the dynamic case, so there's no compatibility problem in making that restriction.

Also, I didn't implement support for wrapped numeric input in this case, because honestly this is enough work already and is just for compatibility.

I still need to document this.  I still need to document UDVTs, too.

gnidan commented on packages/codec/lib/wrap/address.ts:
```diff
@@ -0,0 +1,342 @@
+  ...addressCasesBasic
+];
+
+function* addressFromHexString(
+  dataType: AddressLikeType,
+  input: unknown,
+  wrapOptions: WrapOptions
+): Generator<never, AddressLikeValue, WrapResponse> {
```
I think this might still be worthwhile? Perhaps we can discuss this via screenshare and we can make a decision synchronously.

gnidan replied:
After discussion and head-scratching, seems like leaving it the way it is is, in fact, the best way to go right now.

gnidan commented on packages/codec/lib/wrap/address.ts:
```diff
@@ -0,0 +1,342 @@
+      input,
+      wrapOptions.name,
+      1,
+      "Input was not a 0x-prefixed hex string"
```
Nah, this is probably fine.

gnidan commented on packages/codec/lib/wrap/errors.ts:
```diff
@@ -0,0 +1,108 @@
+ * type it was supposed to be.
+ * @category Errors
+ */
+export class TypeMismatchError extends Error {
```
Nah, we can skip this for now. Sounds like it's not worth the trouble.

gnidan commented on packages/codec/lib/wrap/bool.ts:
```diff
@@ -0,0 +1,240 @@
+>[] = [
+  boolFromString,
+  boolFromBoxedPrimitive,
+  boolFromBoolValue,
+  boolFromBoolError,
```
Hm, sure, that's fine!

haltman-at replied:
Cool, made this change.

haltman-at commented:
One more change: I made sure that the encoder now normalizes any ABI entries that are input (the normalization happens in `getMethod`, since they all end up there eventually).

haltman-at commented:
OK @gnidan, I think it's likely time for rereview on this. :)  There are things I could potentially do before rereview:

1. I didn't add any tests of what happens when contract matching fails (i.e., we now allow it with a default result rather than throwing).  I could add that in.
2. The documentation could be rewritten to focus more on `encodeTransaction`.
3. I could add `forAbi` (but I'm thinking that's probably best left for a separate PR).

So if you want me to do any of those first, I'll do those first... but if you think they can wait, I think I'm going to compile that list of questions for you and then hit "request rereview". :)

haltman-at commented:
OK @gnidan I've put together that list of questions.  So first you can tell me whether to address the three points above, then you can rereview with the following questions in mind.

There are two big categories of questions: The ENS options interface, and questions of what forms of input to accept.

Currently, you can set the ENS settings via `ensSettings`.  *However*, I also made it so that if you pass a provider (for non-ENS purposes), then by default, ENS will be enabled and use that provider; if you want to pass a provider, but not use ENS, you must explicitly disable ENS.  It's not clear to me whether this is a good idea.

In particular, I did this even for `decoder`, not just for `encoder`; and in `decoder`, you *have* to pass a provider!  So ENS will be on by default.  Again, it's not clear to me whether this is a good idea.

(Note that of course I have not yet added reverse ENS resolution to the decoder; currently this ENS will only be used in the decoder for watching mapping keys. :P )

The second category of questions is about what forms of input to accept.  There are any number of forms of input we could accept that I didn't implement.  Some of them I think should likely be added.  I want to get your opinions on which ones to include.  (I'll also say what my opinion is on each.)

1. Should we allow underscores inbetween digits in numeric strings?

I'm leaning yes, on the basis that Solidity allows it (so it goes together with your parser work).  The reason not to is that it's going to be a pain to implement -- it's not so easy to determine which underscores to allow and which ones not to.  But I feel like the argument in favor is compelling enough that I'm inclined to anyway.

2. In `{ type: ..., value: ... }` input, should we allow more type aliases coming from Vyper?  E.g., should we allow `"decimal"` (Vyper's name for it) for `"fixed168x10"`?  I'm leaning yes on that one because it's easier than remembering `"fixed168x10"`; leaning no on allowing more (which I think would basically just be capitalization variants!).

3. Should we allow `Uint8Array` input for strings?  This would be a way of allowing one to input strings with invalid UTF-8 (other than as a `Format.Values.StringValue`).  I'm pretty split on this one.  It seems good to add, but it does have the annoyance that, because we're using the decoder output format, we'd still have to check if the input was valid UTF-8 in order to return something in the proper format.  (Alternatively, should there be some *other* way to allow invalid UTF-8?  It would necessarily have the same problem, though.)

4. Should we allow more units on integer types?  I'm thinking no, it makes things more complicated and carries little benefit that I can see.

5. Should we allow `decimal.js` input? :P  As a reminder, just as `big.js` is the stripped-down version of `bignumber.js`, `decimal.js` is the even-more-heavyweight version.  This seems like an obvious no, but I figured I'd ask. :P

6. When accepting numeric input for dynamic bytes types, should we allow codec-format wrapped numbers?  For consistency the answer should be "yes", but... ugh, I just don't want to do this.  It's more work for a rare case, and the whole "numbers as bytes" thing is just for compatibility anyway.

7. Should we allow base64-encoded bytestring for `bytes` and `bytesN` input?  I think the answer is no, but I want to explain why this is something we might want to do, because there *is* a reason we might want to do this.  (And then I'll explain why actually we shouldn't.)

The reason we might want to is `privateFor`.  Recall, Truffle Encoder doesn't just handle recognition of arguments to functions; it also handles transaction options, and one of those transaction options is `privateFor`.  The `privateFor` option is given as an array of strings; those strings should be each be a base64-encoded bytestring of length 32 (to be clear, that's the length of the bytestring, not the length of the base64 encoding).

The problem is that this introduces an ugly special case into options recognition.  The other transaction options are all either an unsigned integer (corresponding to Solidity `uint`); an address (`address`); a bytestring (`bytes`); or a boolean (`bool`).  So we can use our existing wrap functions for handling these.  (Note that I haven't yet implemented `accessList` in the transaction options in Encoder -- I was waiting on Ganache 7 for that -- but, while its type is more complex, it can still be handled by existing mechanisms; I just need to construct that more complex type and pass it in, then nativize the result afterward.)

But then there's `privateFor`.  Because each string is supposed to be in base64, this *can't* be handled by existing mechanisms.  Instead it's just stuck in the code as an ugly special case.  This raises the question: Should we add base64 as a valid input for bytestring types?  Then, `privateFor` would no longer be as much of a special case, and could use simpler code, like the other cases (OK, more complex than the other cases, but simpler than it currently is).

But, I think this idea has some complications and some problems.  Let me start with a minor complication, which is that this would require being able to decode (and then re-encode to) base64.  Presumably we'd use a library for that.  Currently, we don't actually ever decode base64; we just use regular expressions to determine whether it's valid and if so what length the bytestring it represents is (and then if it's valid and the appropriate length, we pass it through unmodified, without any attempt to decode it).  I don't think this is really a big deal, but it would mean having to go look for an appropriate library.

The real problem to my mind is... what on earth would this do to overload resolution, or to input interpretation more generally?  So many other strings that a user might enter are also going to happen to be valid base64, and if we unexpectedly interpret them that way when the user isn't expecting that, that seems like a potential problem.  Like, don't forget that, depending on its length, a *hex* string is likely going to be valid base64!  Yikes!  (And decimal strings, too...)  I think it should (generally) be obvious how any particular input will be interpreted; and while it's possible that Encoder doesn't always make it totally obvious, I think allowing base64 makes things totally wild in a way that is kind of user-hostile.  (I already don't like how we have to allow entering odd-length hex strings for `bytes` input for compatibility with ethers.)

There is a third option: To add a `base64Allowed` flag to `WrapOptions`, which `wrapTxOptions` could then set.  Maybe that's the way to do it?  I dunno, that honestly sounds to me even hackier than the current hack.

Anyway that's my list of questions!

haltman-at commented:
OK @gnidan, what with #4491 being merged, I've updated this PR to include `type` and `accessList` transaction options, as well as new tests for these.  Some notes on these:

1. Due to [this issue](https://github.com/ChainSafe/web3.js/issues/4597), I had the encoder always output `type` as a hex string, rather than as a BN.
2. I typed `type` as `uint8`.  Of course, types `0xc0` and larger are illegal, but I figure we don't need to check that; checking that can be the responsibility of whatever library is performing the transaction.
3. I typed the storage keys of an access list as `uint256` rather than `bytes32`.  This is because I figured it's much more likely that people would want to use numerical input for these, than that they'd want to use truncated bytestrings.  Note that this technically disagrees with the existing behavior, which is to pad short hex strings on the *right*, not on the left, but, well, we discussed this and agreed this is fine.  Note that on *output* we always ouput a full 32-byte bytestring in order to prevent problems.
4. In order to facilitate said output, I added a special `nativizeAccessList` function.  Because the decoder output format doesn't have as strong typing as I'd like, this is just stuffed full of coercions.  Oh well.  It works, and it's got a comment warning about appropriate use.  Note that in order to avoid weird dependencies, I put this in `format/utils/inspect.ts` rather than `export.ts`; I did however have `export.ts` re-epxort it.
5. Now that `type` is a legal tx option, this does raise the specter of, does this cause problems with type/value input?  I don't really think this should be a problem though... I mean, the values of `type` for each case should be entirely different.

...unrelatedly, I just ran `madge` and found a circular dependency I'll need to go eliminate.  Dammit.  Well, I'll push again once I have.

haltman-at commented:
> I typed type as uint8. Of course, types 0xc0 and larger are illegal, but I figure we don't need to check that; checking that can be the responsibility of whatever library is performing the transaction.

Actually, @gnidan, I'm not entirely certain of this decision, so let me tack this on as one more question for you: **Should we reject values of `type` that are greater than or equal to `0xc0`?**

gnidan commented:
To address your first three points:

> 1. I didn't add any tests of what happens when contract matching fails (i.e., we now allow it with a default result rather than throwing).  I could add that in.

Seems worthwhile to include these tests!

> 2. The documentation could be rewritten to focus more on `encodeTransaction`.

100% yes this seems worth doing

> 3. I could add `forAbi` (but I'm thinking that's probably best left for a separate PR).

Sure, this can be separate.

And to your open questions:

> There are two big categories of questions: The ENS options interface, and questions of what forms of input to accept.
> 
> 1. Currently, you can set the ENS settings via `ensSettings`.  _However_, I also made it so that if you pass a provider (for non-ENS purposes), then by default, ENS will be enabled and use that provider; if you want to pass a provider, but not use ENS, you must explicitly disable ENS.  It's not clear to me whether this is a good idea.

This seems fine. I think we're dealing with edge cases there, so I don't think it's a problem. Did you have a specific reason why it wouldn't be a good idea to force explicitly disabling ENS?

> 2. In particular, I did this even for `decoder`, not just for `encoder`; and in `decoder`, you _have_ to pass a provider!  So ENS will be on by default.  Again, it's not clear to me whether this is a good idea.  (Note that of course I have not yet added reverse ENS resolution to the decoder; currently this ENS will only be used in the decoder for watching mapping keys. :P )

I don't see the problem with ENS being turned on by default!



> The second category of questions is about what forms of input to accept. There are any number of forms of input we could accept that I didn't implement. Some of them I think should likely be added. I want to get your opinions on which ones to include. (I'll also say what my opinion is on each.)
> 
> 3. Should we allow underscores inbetween digits in numeric strings?  I'm leaning yes, on the basis that Solidity allows it (so it goes together with your parser work).  The reason not to is that it's going to be a pain to implement -- it's not so easy to determine which underscores to allow and which ones not to.  But I feel like the argument in favor is compelling enough that I'm inclined to anyway.

Yes! Definitely!

> 4. In `{ type: ..., value: ... }` input, should we allow more type aliases coming from Vyper?  E.g., should we allow `"decimal"` (Vyper's name for it) for `"fixed168x10"`?  I'm leaning yes on that one because it's easier than remembering `"fixed168x10"`; leaning no on allowing more (which I think would basically just be capitalization variants!).

Yes to `"decimal"`; I'm okay stopping there and not adding more (for now)

> 5. Should we allow `Uint8Array` input for strings?  This would be a way of allowing one to input strings with invalid UTF-8 (other than as a `Format.Values.StringValue`).  I'm pretty split on this one.  It seems good to add, but it does have the annoyance that, because we're using the decoder output format, we'd still have to check if the input was valid UTF-8 in order to return something in the proper format.  (Alternatively, should there be some _other_ way to allow invalid UTF-8?  It would necessarily have the same problem, though.)

I don't think this is necessary? Maybe can just wait until someone wants it.

> 6. Should we allow more units on integer types?  I'm thinking no, it makes things more complicated and carries little benefit that I can see.

Nah.

> 7. Should we allow `decimal.js` input? :P  As a reminder, just as `big.js` is the stripped-down version of `bignumber.js`, `decimal.js` is the even-more-heavyweight version.  This seems like an obvious no, but I figured I'd ask. :P

Obvious no :)

> 8. When accepting numeric input for dynamic bytes types, should we allow codec-format wrapped numbers?  For consistency the answer should be "yes", but... ugh, I just don't want to do this.  It's more work for a rare case, and the whole "numbers as bytes" thing is just for compatibility anyway.

I agree with your reasoning (would be nice for consistency, but not worth the effort if it's a pain)

> 9. Should we allow base64-encoded bytestring for `bytes` and `bytesN` input?  I think the answer is no, but I want to explain why this is something we might want to do, because there _is_ a reason we might want to do this.  (And then I'll explain why actually we shouldn't.)

Let's skip this support for now, for the reasons you describe.


> Actually, @gnidan, I'm not entirely certain of this decision, so let me tack this on as one more question for you: **Should we reject values of `type` that are greater than or equal to `0xc0`?**

I think this is fine? I know David brought up a reason that we want to allow it for legacy reasons, but it seems safe to reject.

haltman-at commented:
Thanks Nick!  Just to clarify some things:

> This seems fine. I think we're dealing with edge cases there, so I don't think it's a problem. Did you have a specific reason why it wouldn't be a good idea to force explicitly disabling ENS?

I don't, really, but like the fact that you said that ENS should get its own separate provider option made me think you maybe wanted it off by default.

> I don't think this is necessary? Maybe can just wait until someone wants it.

Cool, I think I'll skip this then.

> > Actually, @gnidan, I'm not entirely certain of this decision, so let me tack this on as one more questn for you: Should we reject values of type that are greater than or equal to 0xc0?

> I think this is fine? I know David brought up a reason that we want to allow it for legacy reasons, but it seems safe to reject.

To be clear, currently encoder *doesn't* reject those; that would be a change.  But it's an easy change. :)

haltman-at commented:
OK, I've added support for the stuff discussed above.

A note about underscores in hex strings: I've allowed it inbetween hex digits within the same byte, even for byte strings for which we might want to disallow it.  This is because Solidity allows it in `0x` literals, so I'm allowing it too.  (It doesn't allow it in `hex"..."` literals, but encoder doesn't have those.)

haltman-at commented:
...also I went and added support for `Uint8Array`s as strings, because why not. :P

gnidan reviewed (approved):
Just went through the docs and played with this. I think this is good to merge :)

haltman-at commented:
OK, well I'm going to go back and split up that test file before actually merging this. :P

haltman-at commented:
OK, I've now split up the tests -- the resulting files are still pretty big, but, it's no longer totally ridiculous.  So, if you want to look any of that over again...

gnidan reviewed (approved):
Test splitup looks nice :)
